services/masterdata/src/lib/components/translation/upsert-translation-form/upsert-translation.component.ts

```diff
@@ -0,0 +1,131 @@
+import { LanguageExistsValidator } from "../../../core/features";
+import { IMasterdataTranslationFlat } from "../../../core/interfaces";
+import { LanguageItem } from "../../../core/models";
+import { MasterdataRendererService } from "../../../core/services";
+import { MasterdataMaxLength, getFieldValue, getRequiredFieldValue } from "../../../core/utils";
+
+@Component({
+	templateUrl: "./upsert-translation.component.html",
+	styleUrls: ["./upsert-translation.component.scss"]
+})
+export class MasterdataTranslationUpsertComponent extends DialogComponent implements OnInit {
+	isLoading = false;
+	saveFormButtonText!: string;
+	dataForm!: FormGroup;
+	isFormSubmitted = false;
+	maxLength = MasterdataMaxLength;
+	item!: IMasterdataTranslationFlat;
+	isTypeModel!: boolean;
+	isAddForm!: boolean;
+	translatableFields: string[] = [];
+	currentLanguages: string[] = [];
+	languagesList: LanguageItem[] = [];
+	availableLanguages: LanguageItem[] = [];
+
+	constructor(
+		private readonly masterdataRenderer: MasterdataRendererService,
+		private readonly languageExistsValidator: LanguageExistsValidator,
+		private readonly languageService: LanguageService,
+		private readonly formBuilder: FormBuilder,
+		@Inject("LensDialogRef") private readonly ref: DialogRef,
+		@Inject("LensDialogConfig") public readonly config: DialogConfig
+	) {
+		super();
+		this.initConfigData(this.config.data);
+	}
+
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	initConfigData(data: any) {
+		const { item, isTypeModel, isAddForm, translatableFields, currentLanguages, languagesList } = data;
+		this.item = item;
+		this.isTypeModel = isTypeModel;
+		this.isAddForm = isAddForm;
+		this.translatableFields = translatableFields;
+		this.currentLanguages = currentLanguages;
+		this.languagesList = languagesList;
+	}
+
+	ngOnInit(): void {
+		this.availableLanguages = this.languagesList.filter(item => !this.currentLanguages.includes(item.code));
+
+		const whenMdItem = !this.isTypeModel && {
+			value: ["", [Validators.required, Validators.maxLength(this.maxLength.value)]]
+		};
+		const whenAddForm = this.isAddForm && {
+			language: ["", [Validators.required, this.languageExistsValidator.checkIfLanguageExists(this.currentLanguages)]]
+		};
+		this.dataForm = this.formBuilder.group({
+			...whenAddForm,
+			isDefault: [false, [Validators.required]],
+			...whenMdItem,
+			name: ["", [Validators.required, Validators.maxLength(this.maxLength.name)]],
+			description: ["", [Validators.maxLength(this.maxLength.description)]]
+		});
+
+		if (!this.isAddForm) {
+			this.populateFormValues(this.item);
+		}
+
+		this.buildTranslationTexts();
+	}
+
+	// convenience getter for easy access to form fields
+	get getFormFields() {
+		return this.dataForm.controls;
+	}
+
+	onCancelButtonClicked() {
+		this.ref.close();
+	}
+
+	onSaveButtonClicked() {
+		this.isFormSubmitted = true;
+		if (this.dataForm.invalid) {
+			// stop here if form is invalid
```

Comment:
This comment is like the "//this is a bridge" sign :D :D
https://i.redd.it/kxeexaamhes51.jpg